### ISSUE TITLE
data store: n=1 tasks should be created with the appropriate runahead status

### ISSUE THREAD

**Issue by oliver-sanders** [cylc/cylc-flow#5580]
Tasks which are beyond the runahead limit are only marked as runahead limited when they are spawned.

Here's a screenshot showing a cycle which is beyond the runahead limit:

![Screenshot from 2023-06-13 12-45-41](https://github.com/cylc/cylc-flow/assets/16705946/22d6b621-d87a-4ef7-a226-b993bb816898)

The `OS_START` tasks show as runahead-limited, whereas the others don't. This is because the `OS_START` tasks are parentless i.e. n=0 so have been spawned, the other tasks are downstreams of `OS_START` so are n>0 tasks generated by the data store.

This is a cause of user confusion as it makes it look like the runahead limit has not taken effect. As per #3748 user's are not expected to understand why a task does or does not exist in the pool. Note it is possible that the GUI may show a runahead limited cycle with no runahead limited tasks (e.g. if viewing the  n=2 window).

So IMO the data store should really be setting (and updating) the `is_runahead` flag so that all tasks beyond the runahead limit are displayed as limited irrespective of internal implementation details (i.e. spawning).

**Comment by hjoliver** [comment]
I'm not sure about this.

Technically, runahead limited means *ready to run, but held back by the runahead limit*. 

So with `foo => bar`, if `foo` is runahead limited, `bar` is not - it is not being held back by the runahead limit, because it is not actually ready to run yet.

> Note it is possible that the GUI may show a runahead limited cycle with no runahead limited tasks (e.g. if viewing the n=2 window).

That cycle is in the future graph, and is not being held back by the runahead limit because the active part of the workflow has not even reached that part of the graph yet.

This is a useful distinction when you can't see the dependencies. If the runahead limit moves on, you can see that `foo` will start running immediately, but `bar` will not.

Should all future tasks be shown as queue limited too? (And what about xtriggers - that's difficult because we'd have to evaluate the trigger to see if it is satisfied, but in principle it's somewhat analagous to runahead limiting).

> As per https://github.com/cylc/cylc-flow/issues/3748 user's are not expected to understand why a task does or does not exist in the pool.

With the tiny exception that `n=0` is special. That isn't just a matter of internal implementation, because the idea of the "active window" on the graph is important to understand.

**Comment by hjoliver** [comment]
> IMO the "technically" bit is really an internal implementation detail,

Technically, that was an unfortunate use of the word "technical" in the colloquial sense, by me!

I guess disagreement here boils down to this: should we put a runahead icon on:
  1. every task that is "beyond the runahead limit" 
OR
  2. every task that is "actively runahead limited" (i.e., it would run immediately if not for the limit). 

(Note 2. is not just parentless tasks; it's any task with satisfied prerequisites as well).

Both of these are conceptual, not just technical implementation.

IMO the distinction does convey useful information. We often want to know why a particular waiting task is not running yet. With 2., we can see exactly which tasks have no unsatisfied prerequisites and will run as soon as the runahead limit advances.

This should not cause user confusion so long as they understand what the runahead limit is, and specifically that it is cycle point based.

Maybe we can have the best of both worlds:
- mark **tasks** that are actively runahead limited
   - including all visible parentless tasks, which we are not doing now
- mark **cycle points** that are beyond the runahead limit
   - this could be the runahead icon, and/or colouring (or greying out) the cycle point, and/or displaying a literal "line in the sand" across the display (possible in the tree view, and the table view when sorted by point,...)

**Comment by oliver-sanders** [comment]
> We often want to know why a particular waiting task is not running yet

This used to be covered by the "show prerequisites and outputs" dialogue, a functionality which will be delivered in the new "metadata view".

We don't visually distinguish a waiting task with satisfied prereqs from a waiting task with unsatisfied prereqs, but in those cases the task will change state imminently so point taken on the runahead limit **'ed** point.

IMO it's more important to convey the runahead limit as opposed to tasks which are runahead limit **'ed** so I'm still leaning towards (1).

> This should not cause user confusion so long as they understand what the runahead limit is, and specifically that it is cycle point based.

I opened this issue because it did cause confusion, both with the user and myself. Taking a look at the above screenshot, it's not really clear that all of those tasks are behind the runahead limit. The user had also filtered tasks so the runahead limited ones weren't showing which made it even more confusing. The cycle point should definitely be marked as runahead, but that still doesn't make it especially obvious and doesn't cover all bases, consider the table and graph views where the cycle point doesn't show and there is absolutely no information anywhere telling you that these tasks are behind the runahead limit.

**Comment by hjoliver** [comment]
> The user had also filtered tasks so the runahead limited ones weren't showing which made it even more confusing

That's a good point.

>  consider the table and graph views where the cycle point doesn't show and there is absolutely no information anywhere telling you that these tasks are behind the runahead limit.

Cycle point does appear in the table view, but I guess sorting on other columns would complicate the runahead display problem a bit.

> This used to be covered by the "show prerequisites and outputs" dialogue, a functionality which will be delivered in the new "metadata view".

Yes, but for this particular problem we can easily convey the relevant information without needing an additional click menu operation.

OK, how about this:
- mark runahead cycle points
- mark all tasks that are beyond the runahead limit, with a badge
- highlight the actively limited tasks (i.e., the only ones currently marked) by emphasizing the badge somehow (colour or contrast...)

**Comment by oliver-sanders** [comment]
Note, historically, if users wanted to know why a task was held back, they used the "show prereqs and outputs" button. In the new GUI, the "metadata view" will provide this information along with task `[meta]`, broadcast info and other useful things. IMO this is how users should differentiate between a task which is beyond the runahead limit and a task which is held back by the runahead limit alone. The metadata view is steadily climbing up the priority ladder, especially the prereqs/outputs part which is currently a glaring omission in Cylc 8.

> mark runahead cycle points

Note, this is a client rather than a scheduler thing, Tui already marks a cycle as runahead if any task within that cycle displayed in the n-window is runahead. This means that the cycle will not be marked as runahead if all tasks within it are held back by at least one dependency.

**Comment by hjoliver** [comment]
OK (as per Element chat thread) - I concede defeat!

Personally I still think it's useful to distinguish:
- tasks that are ready to run as soon as the runahead limit moves on (i.e., actively limited)
- tasks that are downstream of those tasks and therefore have unmet prerequisites as well (i.e., the entire future graph)

However, I accept that marking all such tasks as runahead limited would not be incorrect and it probably would be easier to understand for most users.

**Comment by hjoliver** [comment]
For a clear(er) description of the current situation, at least, see https://github.com/cylc/cylc-doc/pull/824#discussion_r2083736536

**Comment by dwsutherland** [comment]
I'm gonna throw in a different take:

Uses should be able to tell with tasks are active/n=0 and which are not/n>0, at a glance..

I'm not sure what the solution would be.. Maybe if they're slightly faded and/or have an italic look .. Or some other solution.
Regardless of the solution, this would make it clear that these tasks are not like the others.. Not only for the runahead case, but also for when users are trying to take actions on tasks that don't apply to n>0 (i.e. remove)

I'm forever telling my colleague to set `n=0` to find out what tasks are actually holding up a workflow (as he likes to set `n=3` a lot), usually these are old tasks that are waiting on an xtrigger that is not satisfied for some upstream reason..

**Comment by hjoliver** [comment]
@dwsutherland I agree that users should be able to distinguish `n=0`, i.e. "active" tasks, within the n-window without having to select 0-depth only, although that's really a separate issue from how runahead limiting is expressed in the GUI. I'm sure we must have discussed this before, but didn't find it in a quick look through cylc-ui issues - @cylc/core thoughts?

**Comment by dwsutherland** [comment]
Well, I'm not sure it is an entirely separate issue; it means users can clearly see what task matter WRT being stuck/halted runahead or otherwise.

**Comment by oliver-sanders** [comment]
Only displaying the runahead icon for n=0 tasks is not a solution to highlighting the n=0 window (as it only serves as an indicator in niche circumstances) so this conversation is off topic for the current issue.

**Comment by hjoliver** [comment]
Yes, we need to distinguish n=0 regardless of runahead limiting, so it's a separate issue.

But, nevertheless, it came up here, so I just asked if anyone was aware of existing (esp. cylc-ui) issues that cover that- are you aware of any?  If not, I'll post a new one.

**Comment by hjoliver** [comment]
... https://github.com/cylc/cylc-ui/issues/2179

**Comment by dwsutherland** [comment]
> Only displaying the runahead icon for n=0 tasks is not a solution to highlighting the n=0 window (as it only serves as an indicator in niche circumstances) so this conversation is off topic for the current issue.

Ok, let me state it in less general and on topic way:  

`n=1` tasks don't need to be marked as runahead if it's visually clear that they are inactive.
Because only thing it tells you is "these _may_ be runahead blocked if active", but otherwise meaningless.

However I realize you could probably spin it another way, but my main point is active/inactive should be the main distinguishing factor...

**Comment by hjoliver** [comment]
I was initially of your opinion @dwsutherland but I've pretty much come around to @oliver-sanders perspective on this.

I updated the User Guide to explain the status quo better for 8.5, and this came up on the forum the other day too: https://cylc.discourse.group/t/wui-runahead-limit-is-showing-for-the-next-cycle-but-not-the-ones-afterwards/1175/2 ... but as hinted at the bottom of my post, I now think users will find it more useful just to which cycles are below the runahead limit and which are beyond it.  
(For a complete understanding of the system, which occasionally helps figure out what's going on in edge cases, it would be good if users understood the difference between "actively runahead limited n=0 task" and "n>0 task beyond the runahead limit", but the reality is that's way beyond most users).

**Comment by oliver-sanders** [comment]
Completely unrelated to this, but worth noting, we used to filter out runahead tasks in the GUI by default.

TBH, this is probably still a sensible default.
